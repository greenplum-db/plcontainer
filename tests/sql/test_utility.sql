\! $(pwd)/test_utility.sh
