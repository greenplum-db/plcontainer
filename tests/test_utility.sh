#!/bin/bash
#------------------------------------------------------------------------------
#
#
# Copyright (c) 2017-Present Pivotal Software, Inc
#
#------------------------------------------------------------------------------
#
# For plcontainer CLI testing. Called in sql/test_utility.sql

plcontainer -h

######################## test image-add ###############################
echo
echo "Test image-add: negative cases"
plcontainer image-add
plcontainer image-add nonexist
plcontainer image-add -f nonexist_file | sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'
plcontainer image-add -u nonexist_url | sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'

######################## test image-delete ###############################
echo
echo "Test image-delete: negative cases"
plcontainer image-delete
plcontainer image-delete nonexist

######################## test image-list ###############################
echo
echo "Test image-list: negative cases"
plcontainer image-list >/tmp/tmp.image-list.out # We just care about whether there is stderr.
plcontainer image-list not_exist

echo
echo "Prepare a blank runtime configuration file and test runtime-backup"
#Backup the runtime configurations.
plcontainer runtime-backup -f test_backup_cfg_file
cat <<EOF >tmp_cfg_file
<?xml version="1.0" ?>
<configuration>
</configuration>
EOF
plcontainer runtime-restore -f tmp_cfg_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://' \
	| grep -v 'Distributing to'
# We do not want $GPHOME to be in exp output.
plcontainer runtime-backup -f /tmp/backup_file; cat /tmp/backup_file |  sed -e "s|${GPHOME}|GPHOME|"
rm -f tmp_cfg_file

######################## test runtime-add/delte/show/backup ###############################
echo
echo "Test runtime-add: nagative cases"
plcontainer runtime-add -r nonexist_runtime
plcontainer runtime-add -i nonexist_image
plcontainer runtime-add -r nonexist_runtime -i nonexist_image
plcontainer runtime-add -r nonexist_runtime -i nonexist_image -l java

echo
echo "Test runtime-add, runtime-backup, and runtime-delete"
plcontainer runtime-add -r runtime1 -i image1 -l python \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://' \
	| grep -v 'Distributing to'

plcontainer runtime-backup -f /tmp/backup_file; cat /tmp/backup_file |  sed -e "s|${GPHOME}|GPHOME|"
plcontainer runtime-delete -r runtime1 | sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://' \
	| grep -v 'Distributing to'
plcontainer runtime-backup -f /tmp/backup_file; cat /tmp/backup_file |  sed -e "s|${GPHOME}|GPHOME|"
plcontainer runtime-delete -r runtime1 | sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://' \
	| grep -v 'Distributing to'

echo
echo "Test runtime-add with shared directories"
plcontainer runtime-add -r runtime1 -i image1 -l python -v /host_dir/shared:/container_dir/shared:rw \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://' \
	| grep -v 'Distributing to'
plcontainer runtime-delete -r runtime1 \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://' \
	| grep -v 'Distributing to'
plcontainer runtime-add -r runtime1 -i image1 -l python -v /host_dir1/shared:/container_dir1/shared:rw \
		-v /host_dir2/shared:/container_dir2/shared \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://' \
	| grep -v 'Distributing to'
plcontainer runtime-delete -r runtime1 \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://' \
	| grep -v 'Distributing to'

echo
echo "Test runtime-add with both shared directories and settings"
plcontainer runtime-add -r runtime1 -i image1 -l python -v /host_dir1/shared1:/container_dir1/shared1:rw \
		-v /host_dir1/shared2:/container_dir1/shared2:ro \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://' \
	| grep -v 'Distributing to'
plcontainer runtime-add -r runtime2 -i image2 -l r -v /host_dir2/shared1:/container_dir2/shared1:rw \
		-v /host_dir2/shared2:/container_dir2/shared2:ro -s memory_mb=512 -s use_network=yes \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://' \
	| grep -v 'Distributing to'
plcontainer runtime-show -r runtime_not_exist \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'
plcontainer runtime-show -r runtime_not_exist >/dev/null; echo $?
plcontainer runtime-show |  sed -e "s|${GPHOME}|GPHOME|"
plcontainer runtime-show -r runtime1 |  sed -e "s|${GPHOME}|GPHOME|"
plcontainer runtime-delete -r runtime1 \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://' \
	| grep -v 'Distributing to'
plcontainer runtime-delete -r runtime2 \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://' \
	| grep -v 'Distributing to'
plcontainer runtime-show |  sed -e "s|${GPHOME}|GPHOME|" \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://' \

echo
echo "Test runtime-replace: negative cases"
plcontainer runtime-replace -r runtime3
plcontainer runtime-replace -r runtime3 -i image3
plcontainer runtime-replace -r runtime3 -i image3 -l java
plcontainer runtime-replace -r runtime3 -i image3 -l r -s user_network=yes \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'

echo
echo "Test runtime-replace: add a new one"
plcontainer runtime-replace -r runtime3 -i image2 -l r -v /host_dir3/shared1:/container_dir3/shared1:rw \
	-v /host_dir3/shared2:/container_dir3/shared2:ro -s memory_mb=512 -s use_network=yes \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://' \
	| grep -v 'Distributing to'
plcontainer runtime-backup -f /tmp/backup_file; cat /tmp/backup_file |  sed -e "s|${GPHOME}|GPHOME|"
echo
echo "Test runtime-replace: replace"
plcontainer runtime-replace -r runtime3 -i image2 -l r -v /host_dir3/shared3:/container_dir3/shared3:rw \
	-s use_network=no \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://' \
	| grep -v 'Distributing to'
plcontainer runtime-backup -f /tmp/backup_file; cat /tmp/backup_file |  sed -e "s|${GPHOME}|GPHOME|"

echo
echo "Test xml validation: negative cases"

echo
echo "**Test <id> is needed"
cat >bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
    </runtime>
</configuration>
EOF
plcontainer runtime-restore -f bad_xml_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'

echo
echo "**Test: No more than 1 <id> is allowed"
cat >bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python</id>
        <id>plc_python2</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
    </runtime>
</configuration>
EOF
plcontainer runtime-restore -f bad_xml_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'

echo
echo "**Test: <id> naming requirement"
cat >bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>*plc_python</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
    </runtime>
</configuration>
EOF
plcontainer runtime-restore -f bad_xml_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'

echo
echo "**Test <image> is needed"
cat >bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python</id>
        <command>./client</command>
    </runtime>
</configuration>
EOF
plcontainer runtime-restore -f bad_xml_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'

echo
echo "**Test: No more than 1 <image> is allowed"
cat >bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <image>pivotaldata/plcontainer_python:0.2</image>
        <command>./client</command>
    </runtime>
</configuration>
EOF
plcontainer runtime-restore -f bad_xml_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'

echo
echo "**Test: <command> is needed"
cat >bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
    </runtime>
</configuration>
EOF
plcontainer runtime-restore -f bad_xml_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'

echo
echo "**Test: No more than 1 <command> is allowed"
cat >bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <command>./client2</command>
    </runtime>
</configuration>
EOF
plcontainer runtime-restore -f bad_xml_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'

echo
echo "**Test: <shared_directory>: 'container' attr is needed"
cat >bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <shared_directory access="rw" host="/host/plcontainer_clients"/>
    </runtime>
</configuration>
EOF
plcontainer runtime-restore -f bad_xml_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'

echo
echo "**Test: <shared_directory>: 'host' attr is needed"
cat >bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <shared_directory access="rw" container="/clientdir"/>
    </runtime>
</configuration>
EOF
plcontainer runtime-restore -f bad_xml_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'

echo
echo "**Test: <shared_directory>: 'access' attr is is needed"
cat >bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <shared_directory container="/clientdir" host="/host/plcontainer_clients"/>
    </runtime>
</configuration>
EOF
plcontainer runtime-restore -f bad_xml_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'

echo
echo "**Test: <shared_directory>: access must be ro or rw"
cat >bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <shared_directory access="rx" container="/clientdir" host="/host/plcontainer_clients"/>
    </runtime>
</configuration>
EOF
plcontainer runtime-restore -f bad_xml_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'

echo
echo "**Test: <shared_directory>: container paths should not be duplicated"
cat >bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <shared_directory access="rw" container="/clientdir" host="/host/plcontainer_clients"/>
        <shared_directory access="ro" container="/clientdir" host="/host/plcontainer_clients_2"/>
    </runtime>
</configuration>
EOF
plcontainer runtime-restore -f bad_xml_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'

echo
echo "**Test: <shared_directory>: container path should not be /tmp/plcontainer"
cat >bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <shared_directory access="ro" container="/tmp/plcontainer" host="/host/plcontainer_clients_2"/>
    </runtime>
</configuration>
EOF
plcontainer runtime-restore -f bad_xml_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'

echo
echo "**Test: <setting>: must be legal one"
cat >bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <setting logging="Enable"/>
    </runtime>
</configuration>
EOF
plcontainer runtime-restore -f bad_xml_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://' \
	| grep -v 'Distributing to'

echo
echo "**Test: <setting>: memory_mb should be string with integer value"
cat >bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <setting memory_mb="123.4"/>
    </runtime>
</configuration>
EOF
plcontainer runtime-restore -f bad_xml_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'

echo
echo "**Test: <setting>: memory_mb should be string with positive integer value"
cat >bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <setting memory_mb="-123"/>
    </runtime>
</configuration>
EOF
plcontainer runtime-restore -f bad_xml_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'

echo
echo "**Test: <setting>: use_network should be yes or no"
cat >bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <setting use_network="y"/>
    </runtime>
</configuration>
EOF
plcontainer runtime-restore -f bad_xml_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'

echo
echo "**Test: <setting>: logs should be enable or disable"
cat >bad_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <command>./client</command>
        <setting logs="yes"/>
    </runtime>
</configuration>
EOF
plcontainer runtime-restore -f bad_xml_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://'
rm -f bad_xml_file

echo
echo "Test xml validation: postive case"
cat >good_xml_file << EOF
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python</id>
        <image>pivotaldata/plcontainer_python:0.1</image>
        <setting memory_mb="512"/>
        <setting use_network="NO"/>
        <setting logs="Enable"/>
        <command>./client</command>
        <shared_directory access="rw" container="/clientdir" host="/host/plcontainer_clients"/>
    </runtime>
</configuration>
EOF
plcontainer runtime-restore -f good_xml_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://' \
	| grep -v 'Distributing to'
rm -f good_xml_file

#recover the original runtime configurations.
echo
echo "Recover the previous runtime configuration file"
plcontainer runtime-restore -f test_backup_cfg_file \
	| sed -e 's/.*ERROR]://' -e 's/.*INFO]://' -e 's/.*CRITICAL]://' \
	| grep -v 'Distributing to'
rm -f /tmp/tmp.image-list.out /tmp/backup_file
# We do not remove test_backup_cfg_file for checking in case there are test errors.
