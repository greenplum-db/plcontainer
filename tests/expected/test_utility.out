\! $(pwd)/test_utility.sh
usage: plcontainer [-h] [--verbose]
                   {image-add,image-delete,image-list,runtime-add,runtime-replace,runtime-show,runtime-delete,runtime-edit,runtime-backup,runtime-restore,runtime-verify}
                   ...

positional arguments:
  {image-add,image-delete,image-list,runtime-add,runtime-replace,runtime-show,runtime-delete,runtime-edit,runtime-backup,runtime-restore,runtime-verify}
    image-add           Install specified docker image on all hosts
    image-delete        Delete specified docker image on all hosts
    image-list          Show existing container images on the current host
    runtime-add         Add a container runtime configuration on all hosts
    runtime-replace     Replace with a container runtime configuration on all
                        hosts
    runtime-show        Show container runtime configurations on the current
                        host
    runtime-delete      Delete a given container runtime configuration on all
                        hosts
    runtime-edit        Use editor to edit and then update configurations on
                        all hosts
    runtime-backup      Backup the local runtime configurations to a file
    runtime-restore     Restore the runtime configurations from a file and
                        distribute them to all hosts
    runtime-verify      Sanity check configurations on all the hosts

optional arguments:
  -h, --help            show this help message and exit
  --verbose             Enable verbose logging

Test image-add: negative cases
usage: plcontainer image-add [-h] (-f FILE | -u URL)
plcontainer image-add: error: one of the arguments -f/--file -u/--url is required
usage: plcontainer image-add [-h] (-f FILE | -u URL)
plcontainer image-add: error: one of the arguments -f/--file -u/--url is required
-docker image file nonexist_file does not exist
-Getting docker image from nonexist_url
-Cannot pull docker image from nonexist_url

Test image-delete: negative cases
usage: plcontainer image-delete [-h] -i IMAGE_NAME
plcontainer image-delete: error: argument -i/--image is required
usage: plcontainer image-delete [-h] -i IMAGE_NAME
plcontainer image-delete: error: argument -i/--image is required

Test image-list: negative cases
usage: plcontainer [-h] [--verbose]
                   {image-add,image-delete,image-list,runtime-add,runtime-replace,runtime-show,runtime-delete,runtime-edit,runtime-backup,runtime-restore,runtime-verify}
                   ...
plcontainer: error: unrecognized arguments: not_exist

Prepare a blank runtime configuration file and test runtime-backup
-Distributing file plcontainer_configuration.xml to all locations...
-Configuration has changed. Run "select * from plcontainer_refresh_config" in open sessions. New sessions will get new configuration automatically.
<?xml version="1.0" ?>
<configuration>
</configuration>


Test runtime-add: nagative cases
usage: plcontainer runtime-add [-h] -r RUNTIME_ID -i IMAGE_NAME -l {r,python}
                               [-v SHARED] [-s SETTING]
plcontainer runtime-add: error: argument -i/--image is required
usage: plcontainer runtime-add [-h] -r RUNTIME_ID -i IMAGE_NAME -l {r,python}
                               [-v SHARED] [-s SETTING]
plcontainer runtime-add: error: argument -r/--runtime is required
usage: plcontainer runtime-add [-h] -r RUNTIME_ID -i IMAGE_NAME -l {r,python}
                               [-v SHARED] [-s SETTING]
plcontainer runtime-add: error: argument -l/--language is required
usage: plcontainer runtime-add [-h] -r RUNTIME_ID -i IMAGE_NAME -l {r,python}
                               [-v SHARED] [-s SETTING]
plcontainer runtime-add: error: argument -l/--language: invalid choice: 'java' (choose from 'r', 'python')

Test runtime-add, runtime-backup, and runtime-delete
-Distributing file plcontainer_configuration.xml to all locations...
-Configuration has changed. Run "select * from plcontainer_refresh_config" in open sessions. New sessions will get new configuration automatically.
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>runtime1</id>
        <image>image1</image>
        <command>/clientdir/pyclient.sh</command>
        <shared_directory access="ro" container="/clientdir" host="GPHOME/bin/plcontainer_clients"/>
    </runtime>
</configuration>
-Distributing file plcontainer_configuration.xml to all locations...
-Configuration has changed. Run "select * from plcontainer_refresh_config" in open sessions. New sessions will get new configuration automatically.
<?xml version="1.0" ?>
<configuration>
    </configuration>
-This runtime configuraiton does not exist. Ignoring.

Test runtime-add with shared directories
-Distributing file plcontainer_configuration.xml to all locations...
-Configuration has changed. Run "select * from plcontainer_refresh_config" in open sessions. New sessions will get new configuration automatically.
-Distributing file plcontainer_configuration.xml to all locations...
-Configuration has changed. Run "select * from plcontainer_refresh_config" in open sessions. New sessions will get new configuration automatically.
-Distributing file plcontainer_configuration.xml to all locations...
-Configuration has changed. Run "select * from plcontainer_refresh_config" in open sessions. New sessions will get new configuration automatically.
-Distributing file plcontainer_configuration.xml to all locations...
-Configuration has changed. Run "select * from plcontainer_refresh_config" in open sessions. New sessions will get new configuration automatically.

Test runtime-add with both shared directories and settings
-Distributing file plcontainer_configuration.xml to all locations...
-Configuration has changed. Run "select * from plcontainer_refresh_config" in open sessions. New sessions will get new configuration automatically.
-Distributing file plcontainer_configuration.xml to all locations...
-Configuration has changed. Run "select * from plcontainer_refresh_config" in open sessions. New sessions will get new configuration automatically.
PL/Container Runtime Configuration: 
-XML file does not conform XML specification or no related element was found.
-No runtime or no specified runtime was found.
-plcontainer failed. (Reason='runtime-show error') exiting...
1
PL/Container Runtime Configuration: 
---------------------------------------------------------
  Runtime ID: runtime1
  Linked Docker Image: image1
  Runtime Setting(s): 
  Shared Directory: 
  ---- Shared Directory From HOST 'GPHOME/bin/plcontainer_clients' to Container '/clientdir', access mode is 'ro'
  ---- Shared Directory From HOST '/host_dir1/shared1' to Container '/container_dir1/shared1', access mode is 'rw'
  ---- Shared Directory From HOST '/host_dir1/shared2' to Container '/container_dir1/shared2', access mode is 'ro'
---------------------------------------------------------

---------------------------------------------------------
  Runtime ID: runtime2
  Linked Docker Image: image2
  Runtime Setting(s): 
  ---- Container Memory Limited: 512 MB
  ---- Container Use Network For Communication: yes
  Shared Directory: 
  ---- Shared Directory From HOST 'GPHOME/bin/plcontainer_clients' to Container '/clientdir', access mode is 'ro'
  ---- Shared Directory From HOST '/host_dir2/shared1' to Container '/container_dir2/shared1', access mode is 'rw'
  ---- Shared Directory From HOST '/host_dir2/shared2' to Container '/container_dir2/shared2', access mode is 'ro'
---------------------------------------------------------

PL/Container Runtime Configuration: 
---------------------------------------------------------
  Runtime ID: runtime1
  Linked Docker Image: image1
  Runtime Setting(s): 
  Shared Directory: 
  ---- Shared Directory From HOST 'GPHOME/bin/plcontainer_clients' to Container '/clientdir', access mode is 'ro'
  ---- Shared Directory From HOST '/host_dir1/shared1' to Container '/container_dir1/shared1', access mode is 'rw'
  ---- Shared Directory From HOST '/host_dir1/shared2' to Container '/container_dir1/shared2', access mode is 'ro'
---------------------------------------------------------

-Distributing file plcontainer_configuration.xml to all locations...
-Configuration has changed. Run "select * from plcontainer_refresh_config" in open sessions. New sessions will get new configuration automatically.
-Distributing file plcontainer_configuration.xml to all locations...
-Configuration has changed. Run "select * from plcontainer_refresh_config" in open sessions. New sessions will get new configuration automatically.
PL/Container Runtime Configuration: 
-XML file does not conform XML specification or no related element was found.
-No runtime or no specified runtime was found.
-plcontainer failed. (Reason='runtime-show error') exiting...

Test runtime-replace: negative cases
usage: plcontainer runtime-replace [-h] -r RUNTIME_ID -i IMAGE_NAME -l
                                   {r,python} [-v SHARED] [-s SETTING]
plcontainer runtime-replace: error: argument -i/--image is required
usage: plcontainer runtime-replace [-h] -r RUNTIME_ID -i IMAGE_NAME -l
                                   {r,python} [-v SHARED] [-s SETTING]
plcontainer runtime-replace: error: argument -l/--language is required
usage: plcontainer runtime-replace [-h] -r RUNTIME_ID -i IMAGE_NAME -l
                                   {r,python} [-v SHARED] [-s SETTING]
plcontainer runtime-replace: error: argument -l/--language: invalid choice: 'java' (choose from 'r', 'python')
-plcontainer failed. (Reason='Bad setting key: user_network') exiting...

Test runtime-replace: add a new one
-runtime configuration does not exist. Adding a new one.
-Distributing file plcontainer_configuration.xml to all locations...
-Configuration has changed. Run "select * from plcontainer_refresh_config" in open sessions. New sessions will get new configuration automatically.
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>runtime3</id>
        <image>image2</image>
        <command>/clientdir/rclient.sh</command>
        <shared_directory access="ro" container="/clientdir" host="GPHOME/bin/plcontainer_clients"/>
        <shared_directory access="rw" container="/container_dir3/shared1" host="/host_dir3/shared1"/>
        <shared_directory access="ro" container="/container_dir3/shared2" host="/host_dir3/shared2"/>
        <setting memory_mb="512"/>
        <setting use_network="yes"/>
    </runtime>
</configuration>

Test runtime-replace: replace
-Distributing file plcontainer_configuration.xml to all locations...
-Configuration has changed. Run "select * from plcontainer_refresh_config" in open sessions. New sessions will get new configuration automatically.
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>runtime3</id>
        <image>image2</image>
        <command>/clientdir/rclient.sh</command>
        <shared_directory access="ro" container="/clientdir" host="GPHOME/bin/plcontainer_clients"/>
        <shared_directory access="rw" container="/container_dir3/shared3" host="/host_dir3/shared3"/>
        <setting use_network="no"/>
    </runtime>
</configuration>

Test xml validation: negative cases

**Test <id> is needed
-Lack of 'id' subelement in a runtime element.
-XML file does not conform XML specification or wrong settings.
-Validation failed
-Bad xml format. Please fix the file at first.

**Test: No more than 1 <id> is allowed
-Found 2 (1 is legal) 'id' subelements in a runtime element. The first one 'id': plc_python
-XML file does not conform XML specification or wrong settings.
-Validation failed
-Bad xml format. Please fix the file at first.

**Test: <id> naming requirement
-'id' name with wrong format: *plc_python
-XML file does not conform XML specification or wrong settings.
-Validation failed
-Bad xml format. Please fix the file at first.

**Test <image> is needed
-Lack of 'image' subelement in a runtime element in runtime plc_python.
-XML file does not conform XML specification or wrong settings.
-Validation failed
-Bad xml format. Please fix the file at first.

**Test: No more than 1 <image> is allowed
-Found 2 (1 is legal) 'image' subelements in a runtime plc_python. The first one 'image': pivotaldata/plcontainer_python:0.1
-XML file does not conform XML specification or wrong settings.
-Validation failed
-Bad xml format. Please fix the file at first.

**Test: <command> is needed
-Lack of 'command' subelement in a runtime element in runtime plc_python.
-XML file does not conform XML specification or wrong settings.
-Validation failed
-Bad xml format. Please fix the file at first.

**Test: No more than 1 <command> is allowed
-Found 2 (1 is legal) 'command' subelements in runtime plc_python. The first one 'command': ./client
-XML file does not conform XML specification or wrong settings.
-Validation failed
-Bad xml format. Please fix the file at first.

**Test: <shared_directory>: 'container' attr is needed
-Unknown or missing attributes (need and only need 'access', 'host' and 'container') in shared_directory in runtime plc_python
-Current: ['access', 'host']
-XML file does not conform XML specification or wrong settings.
-Validation failed
-Bad xml format. Please fix the file at first.

**Test: <shared_directory>: 'host' attr is needed
-Unknown or missing attributes (need and only need 'access', 'host' and 'container') in shared_directory in runtime plc_python
-Current: ['access', 'container']
-XML file does not conform XML specification or wrong settings.
-Validation failed
-Bad xml format. Please fix the file at first.

**Test: <shared_directory>: 'access' attr is is needed
-Unknown or missing attributes (need and only need 'access', 'host' and 'container') in shared_directory in runtime plc_python
-Current: ['host', 'container']
-XML file does not conform XML specification or wrong settings.
-Validation failed
-Bad xml format. Please fix the file at first.

**Test: <shared_directory>: access must be ro or rw
-access should be 'ro' or 'rw' in shared_directory in runtime plc_python, but now: rx
-XML file does not conform XML specification or wrong settings.
-Validation failed
-Bad xml format. Please fix the file at first.

**Test: <shared_directory>: container paths should not be duplicated
-container should not have same mount points (check shared_directory) in runtime plc_python
-Note if you are modifying/adding runtime configuration please do not use the reserved container paths /clientdir and /tmp/plcontainer
-XML file does not conform XML specification or wrong settings.
-Validation failed
-Bad xml format. Please fix the file at first.

**Test: <shared_directory>: container path should not be /tmp/plcontainer
-container path should not be /tmp/plcontainer (reserved for unix domain socket connection) in runtime plc_python
-XML file does not conform XML specification or wrong settings.
-Validation failed
-Bad xml format. Please fix the file at first.

**Test: <setting>: must be legal one
-Found illegal setting in runtime plc_python: {'logging': 'Enable'}
-Distributing file plcontainer_configuration.xml to all locations...
-Configuration has changed. Run "select * from plcontainer_refresh_config" in open sessions. New sessions will get new configuration automatically.

**Test: <setting>: memory_mb should be string with integer value
-memory_mb should be a positive integer in runtime plc_python, but now: '123.4'
-XML file does not conform XML specification or wrong settings.
-Validation failed
-Bad xml format. Please fix the file at first.

**Test: <setting>: memory_mb should be string with positive integer value
-memory_mb should > 0 in runtime plc_python, but now: '-123'
-XML file does not conform XML specification or wrong settings.
-Validation failed
-Bad xml format. Please fix the file at first.

**Test: <setting>: use_network should be yes or no
-use_network should be 'yes' or 'no' in runtime plc_python, but now: 'y'
-XML file does not conform XML specification or wrong settings.
-Validation failed
-Bad xml format. Please fix the file at first.

**Test: <setting>: logs should be enable or disable
-logs should be 'enable' or 'disable' in runtime plc_python, but now: 'yes'
-XML file does not conform XML specification or wrong settings.
-Validation failed
-Bad xml format. Please fix the file at first.

Test xml validation: postive case
-Distributing file plcontainer_configuration.xml to all locations...
-Configuration has changed. Run "select * from plcontainer_refresh_config" in open sessions. New sessions will get new configuration automatically.

Recover the previous runtime configuration file
-Distributing file plcontainer_configuration.xml to all locations...
-Configuration has changed. Run "select * from plcontainer_refresh_config" in open sessions. New sessions will get new configuration automatically.
